# Import Libraries

In [2]:
import pandas as pd
import numpy as np
import datetime as dt

import warnings
warnings.filterwarnings('ignore') # to ignore all of the warnings when run the code

In [3]:
# to know where this file is save
import os
os.getcwd()

'/Users/salsabilafcr/Documents/DATA ANALYTICS/Projects/Python Project/Project'

# Import Dataset

In [4]:
df = pd.read_csv('/Users/salsabilafcr/Documents/DATA ANALYTICS/Projects/Python Project/Data/Online Retail Data.csv', header = 0)
df

order_id product_code                     product_name  quantity  \
0        493410      TEST001          This is a test product.         5   
1       C493411        21539          RETRO SPOTS BUTTER DISH        -1   
2        493412      TEST001          This is a test product.         5   
3        493413        21724  PANDA AND BUNNIES STICKER SHEET         1   
4        493413        84578   ELEPHANT TOY WITH BLUE T-SHIRT         1   
...         ...          ...                              ...       ...   
461768   539991        21618   4 WILDFLOWER BOTANICAL CANDLES         1   
461769   539991        72741            GRAND CHOCOLATECANDLE         4   
461770   539992        21470    FLOWER VINE RAFFIA FOOD COVER         1   
461771   539992        22258          FELT FARM ANIMAL RABBIT         1   
461772   539992        21155            RED RETROSPOT PEG BAG         1   

                 order_date  price  customer_id  
0       2010-01-04 09:24:00   4.50      12346.0  
1       2010-01-04 09:43:00   4.25      14590.0  
2       2010-01-04 09:53:00   4.50      12346.0  
3       2010-01-04 09:54:00   0.85          NaN  
4       2010-01-04 09:54:00   3.75          NaN  
...                     ...    ...          ...  
461768  2010-12-23 16:49:00   1.25          NaN  
461769  2010-12-23 16:49:00   1.45          NaN  
461770  2010-12-23 17:41:00   3.75          NaN  
461771  2010-12-23 17:41:00   1.25          NaN  
461772  2010-12-23 17:41:00   2.10          NaN  

[461773 rows x 7 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461773 entries, 0 to 461772
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      461773 non-null  object 
 1   product_code  461773 non-null  object 
 2   product_name  459055 non-null  object 
 3   quantity      461773 non-null  int64  
 4   order_date    461773 non-null  object 
 5   price         461773 non-null  float64
 6   customer_id   360853 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 24.7+ MB


# Data Cleaning

In [6]:
df_clean = df.copy()

In [7]:
# create date column
df_clean['date'] = pd.to_datetime(df_clean['order_date']).dt.date.astype('datetime64[ns]')
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461773 entries, 0 to 461772
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      461773 non-null  object        
 1   product_code  461773 non-null  object        
 2   product_name  459055 non-null  object        
 3   quantity      461773 non-null  int64         
 4   order_date    461773 non-null  object        
 5   price         461773 non-null  float64       
 6   customer_id   360853 non-null  float64       
 7   date          461773 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 28.2+ MB


In [8]:
# remove all rows without customer_id
df_clean = df_clean[~df_clean['customer_id'].isna()]
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 360853 entries, 0 to 461744
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      360853 non-null  object        
 1   product_code  360853 non-null  object        
 2   product_name  360853 non-null  object        
 3   quantity      360853 non-null  int64         
 4   order_date    360853 non-null  object        
 5   price         360853 non-null  float64       
 6   customer_id   360853 non-null  float64       
 7   date          360853 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 24.8+ MB


In [9]:
# convert customer_id to string
df_clean['customer_id'] = df_clean['customer_id'].astype(str)
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 360853 entries, 0 to 461744
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      360853 non-null  object        
 1   product_code  360853 non-null  object        
 2   product_name  360853 non-null  object        
 3   quantity      360853 non-null  int64         
 4   order_date    360853 non-null  object        
 5   price         360853 non-null  float64       
 6   customer_id   360853 non-null  object        
 7   date          360853 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 24.8+ MB


In [10]:
# make all product_names lowercase
df_clean['product_name'] = df_clean['product_name'].str.lower()
df_clean.head(3)

order_id product_code             product_name  quantity  \
0   493410      TEST001  this is a test product.         5   
1  C493411        21539  retro spots butter dish        -1   
2   493412      TEST001  this is a test product.         5   

            order_date  price customer_id       date  
0  2010-01-04 09:24:00   4.50     12346.0 2010-01-04  
1  2010-01-04 09:43:00   4.25     14590.0 2010-01-04  
2  2010-01-04 09:53:00   4.50     12346.0 2010-01-04

In [11]:
# remove all rows with product_code or product_name contains 'test'
df_clean = df_clean[~(df_clean['product_code'].str.lower().str.contains('test')) | 
                    ~(df_clean['product_name'].str.contains('test'))]
df_clean

order_id product_code                       product_name  quantity  \
1       C493411        21539            retro spots butter dish        -1   
6        493414        21844                     retro spot mug        36   
7        493414        21533          retro spot large milk jug        12   
8        493414        37508    new england ceramic cake server         2   
9        493414       35001G               hand open shape gold         2   
...         ...          ...                                ...       ...   
461740   539988        84380  set of 3 butterfly cookie cutters         1   
461741   539988       84849D              hot baths soap holder         1   
461742   539988       84849B             fairy soap soap holder         1   
461743   539988        22854        cream sweetheart egg holder         2   
461744   539988       47559B                tea time oven glove         2   

                 order_date  price customer_id       date  
1       2010-01-04 09:43:00   4.25     14590.0 2010-01-04  
6       2010-01-04 10:28:00   2.55     14590.0 2010-01-04  
7       2010-01-04 10:28:00   4.25     14590.0 2010-01-04  
8       2010-01-04 10:28:00   2.55     14590.0 2010-01-04  
9       2010-01-04 10:28:00   4.25     14590.0 2010-01-04  
...                     ...    ...         ...        ...  
461740  2010-12-23 16:06:00   1.25     18116.0 2010-12-23  
461741  2010-12-23 16:06:00   1.69     18116.0 2010-12-23  
461742  2010-12-23 16:06:00   1.69     18116.0 2010-12-23  
461743  2010-12-23 16:06:00   4.95     18116.0 2010-12-23  
461744  2010-12-23 16:06:00   1.25     18116.0 2010-12-23  

[360842 rows x 8 columns]

In [12]:
# remove rows with cancelled status, i.e. those whose order_id begins with 'C'
df_clean = df_clean[df_clean['order_id'].str[:1] !='C']
df_clean.head(3)

order_id product_code                     product_name  quantity  \
6   493414        21844                   retro spot mug        36   
7   493414        21533        retro spot large milk jug        12   
8   493414        37508  new england ceramic cake server         2   

            order_date  price customer_id       date  
6  2010-01-04 10:28:00   2.55     14590.0 2010-01-04  
7  2010-01-04 10:28:00   4.25     14590.0 2010-01-04  
8  2010-01-04 10:28:00   2.55     14590.0 2010-01-04

In [13]:
# change the negative quantity value to positive because the negative value only indicates 
# that the order is canceled.
df_clean['quantity'] = df_clean['quantity'].abs()
df_clean.sample(3)

order_id product_code                 product_name  quantity  \
132924   507359        22509  sewing box retrospot design         1   
17947    495148        22189      cream heart card holder         8   
336271   528644        22551     plasters in tin spaceboy         7   

                 order_date  price customer_id       date  
132924  2010-05-07 15:33:00  16.95     17028.0 2010-05-07  
17947   2010-01-21 11:41:00   3.95     16735.0 2010-01-21  
336271  2010-10-24 12:00:00   1.65     14553.0 2010-10-24

In [14]:
# remove rows with negative price
df_clean = df_clean[~(df_clean['price']<0)]
df_clean.sample(3)

order_id product_code                       product_name  quantity  \
325255   527483        22733  3d traditional christmas stickers         4   
212859   516165        16237               sleeping cat erasers         4   
5646     494021        21242                   red spotty plate         8   

                 order_date  price customer_id       date  
325255  2010-10-18 11:38:00   1.25     13700.0 2010-10-18  
212859  2010-07-18 12:20:00   0.21     13566.0 2010-07-18  
5646    2010-01-11 09:53:00   1.69     15738.0 2010-01-11

In [15]:
# creates an amount value, which is the multiplication of quantity and price
df_clean['amount'] = df_clean['quantity'] * df_clean['price']
df_clean.head(3)

order_id product_code                     product_name  quantity  \
6   493414        21844                   retro spot mug        36   
7   493414        21533        retro spot large milk jug        12   
8   493414        37508  new england ceramic cake server         2   

            order_date  price customer_id       date  amount  
6  2010-01-04 10:28:00   2.55     14590.0 2010-01-04    91.8  
7  2010-01-04 10:28:00   4.25     14590.0 2010-01-04    51.0  
8  2010-01-04 10:28:00   2.55     14590.0 2010-01-04     5.1

In [16]:
# replace the product_name of a product_code that has multiple product_names 
# with one of its most frequently occurring product_names
most_freq_product_name = df_clean.groupby(['product_code','product_name'], as_index=False).agg(order_cnt=('order_id','nunique')).sort_values(['product_code','order_cnt'], ascending=[True,False])
most_freq_product_name['rank'] = most_freq_product_name.groupby('product_code')['order_cnt'].rank(method='first', ascending=False)
most_freq_product_name = most_freq_product_name[most_freq_product_name['rank']==1].drop(columns=['order_cnt','rank'])
df_clean = df_clean.merge(most_freq_product_name.rename(columns={'product_name':'most_freq_product_name'}), how='left', on='product_code')
df_clean['product_name'] = df_clean['most_freq_product_name']
df_clean = df_clean.drop(columns='most_freq_product_name')
df_clean.sample(3)

order_id product_code              product_name  quantity  \
162116   516106        21790        vintage snap cards        12   
2448     493865        20725      lunch bag red spotty        10   
346864   538817        22700  black and white dog bowl        24   

                 order_date  price customer_id       date  amount  
162116  2010-07-16 12:42:00   0.85     17683.0 2010-07-16    10.2  
2448    2010-01-07 16:57:00   1.65     15808.0 2010-01-07    16.5  
346864  2010-12-14 12:38:00   0.85     16191.0 2010-12-14    20.4

In [17]:
# remove outlier
from scipy import stats
df_clean = df_clean[(np.abs(stats.zscore(df_clean[['quantity','amount']]))<3).all(axis=1)]
df_clean = df_clean.reset_index(drop=True)
df_clean.head(3)

order_id product_code                     product_name  quantity  \
0   493414        21844                red retrospot mug        36   
1   493414        21533        retro spot large milk jug        12   
2   493414        37508  new england ceramic cake server         2   

            order_date  price customer_id       date  amount  
0  2010-01-04 10:28:00   2.55     14590.0 2010-01-04    91.8  
1  2010-01-04 10:28:00   4.25     14590.0 2010-01-04    51.0  
2  2010-01-04 10:28:00   2.55     14590.0 2010-01-04     5.1

In [18]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350105 entries, 0 to 350104
Data columns (total 9 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      350105 non-null  object        
 1   product_code  350105 non-null  object        
 2   product_name  350105 non-null  object        
 3   quantity      350105 non-null  int64         
 4   order_date    350105 non-null  object        
 5   price         350105 non-null  float64       
 6   customer_id   350105 non-null  object        
 7   date          350105 non-null  datetime64[ns]
 8   amount        350105 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(5)
memory usage: 24.0+ MB


# Preparing Basket Data

In [19]:
basket = pd.pivot_table(df_clean, index='order_id', columns='product_name', values='product_code',
                       aggfunc='nunique', fill_value=0)
basket

product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                             0                                  0   
493427                             0                                  0   
493428                             0                                  0   
493432                             0                                  0   
493433                             0                                  0   
...                              ...                                ...   
539981                             0                                  0   
539982                             0                                  0   
539985                             0                                  0   
539987                             0                                  0   
539988                             0                                  0   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                                 0                          0   
493427                                 0                          0   
493428                                 0                          0   
493432                                 0                          0   
493433                                 0                          0   
...                                  ...                        ...   
539981                                 0                          0   
539982                                 0                          0   
539985                                 0                          0   
539987                                 0                          0   
539988                                 0                          0   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                                0                                 0   
493427                                0                                 0   
493428                                0                                 0   
493432                                0                                 0   
493433                                0                                 0   
...                                 ...                               ...   
539981                                0                                 0   
539982                                0                                 0   
539985                                0                                 0   
539987                                0                                 0   
539988                                0                                 0   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                      0                       0   
493427                                      0                       0   
493428                                      0                       0   
493432                                      0                       0   
493433                                      0                       0   
...                                       ...                     ...   
539981                                      0                       0   
539982                                      0                       0   
539985                                      0                       0   
539987                                      0                       0   
539988                                      0                       0   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [20]:
basket.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16273 entries, 493414 to 539988
Columns: 3842 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: int64(3842)
memory usage: 477.1+ MB


First, create a basket dataframe, to find out what products have been sold.

In [22]:
def encode(x):
    if x == 0:
        return False
    if x > 0:
        return True
basket_encode = basket.applymap(encode)
basket_encode

product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                         False                              False   
493427                         False                              False   
493428                         False                              False   
493432                         False                              False   
493433                         False                              False   
...                              ...                                ...   
539981                         False                              False   
539982                         False                              False   
539985                         False                              False   
539987                         False                              False   
539988                         False                              False   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                             False                      False   
493427                             False                      False   
493428                             False                      False   
493432                             False                      False   
493433                             False                      False   
...                                  ...                        ...   
539981                             False                      False   
539982                             False                      False   
539985                             False                      False   
539987                             False                      False   
539988                             False                      False   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                            False                             False   
493427                            False                             False   
493428                            False                             False   
493432                            False                             False   
493433                            False                             False   
...                                 ...                               ...   
539981                            False                             False   
539982                            False                             False   
539985                            False                             False   
539987                            False                             False   
539988                            False                             False   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                  False                   False   
493427                                  False                   False   
493428                                  False                   False   
493432                                  False                   False   
493433                                  False                   False   
...                                       ...                     ...   
539981                                  False                   False   
539982                                  False                   False   
539985                                  False                   False   
539987                                  False                   False   
539988                                  False                   False   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

Encode DataFrame basket with value True for all values above 0 and False for all values 0.

In [24]:
basket_filter = basket_encode[(basket_encode>0).sum(axis=1)>1]
basket_filter

product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                         False                              False   
493427                         False                              False   
493428                         False                              False   
493432                         False                              False   
493433                         False                              False   
...                              ...                                ...   
539978                         False                              False   
539981                         False                              False   
539982                         False                              False   
539985                         False                              False   
539988                         False                              False   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                             False                      False   
493427                             False                      False   
493428                             False                      False   
493432                             False                      False   
493433                             False                      False   
...                                  ...                        ...   
539978                             False                      False   
539981                             False                      False   
539982                             False                      False   
539985                             False                      False   
539988                             False                      False   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                            False                             False   
493427                            False                             False   
493428                            False                             False   
493432                            False                             False   
493433                            False                             False   
...                                 ...                               ...   
539978                            False                             False   
539981                            False                             False   
539982                            False                             False   
539985                            False                             False   
539988                            False                             False   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                  False                   False   
493427                                  False                   False   
493428                                  False                   False   
493432                                  False                   False   
493433                                  False                   False   
...                                       ...                     ...   
539978                                  False                   False   
539981                                  False                   False   
539982                                  False                   False   
539985                                  False                   False   
539988                                  False                   False   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

Take a transaction with more than one unique product. Because the objective of this analysis is to find out the relationship between one product and another in a transaction. Then, the code will take the rows that have a value greater than 0 or True and sum them to the right (row-wise), to find out how many total orders there are per each transaction.

# Applying Apriori Alogarithm

In [28]:
pip install mlxtend

  Obtaining dependency information for mlxtend from https://files.pythonhosted.org/packages/1c/07/512f6a780239ad6ce06ce2aa7b4067583f5ddcfc7703a964a082c706a070/mlxtend-0.23.1-py3-none-any.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 5.2 MB/s eta 0:00:00a 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [30]:
from mlxtend.frequent_patterns import apriori

frequent_itemset = apriori(basket_filter, min_support=.01, use_colnames=True).sort_values('support', ascending=False).reset_index(drop=True)
frequent_itemset['product_cnt'] = frequent_itemset['itemsets'].apply(lambda x: len(x))
frequent_itemset

support                                           itemsets  product_cnt
0     0.177953               (white hanging heart t-light holder)            1
1     0.099440                         (regency cakestand 3 tier)            1
2     0.096841                          (jumbo bag red retrospot)            1
3     0.079912                 (pack of 72 retro spot cake cases)            1
4     0.078512                    (assorted colour bird ornament)            1
...        ...                                                ...          ...
1189  0.010064  (key fob , back door , key fob , shed, key fob...            3
1190  0.010064                            (wrap,suki and friends)            1
1191  0.010064                    (retrospot padded seat cushion)            1
1192  0.010064  (pack of 6 pannetone gift boxes, pack of 6 bir...            2
1193  0.010064  (jumbo storage bag suki, lunch bag red spotty,...            3

[1194 rows x 3 columns]

List frequent itemsets (a collection of products that are often purchased).

In [34]:
from mlxtend.frequent_patterns import association_rules

product_association = association_rules(frequent_itemset, metric='confidence', 
                                        min_threshold=.7).sort_values(['support','confidence'], 
                                                                      ascending=[False,False]).reset_index(drop=True)
product_association

antecedents  \
0                  (red hanging heart t-light holder)   
1                    (sweetheart ceramic trinket box)   
2                                 (toilet metal sign)   
3                      (red retrospot sugar jam bowl)   
4                      (painted metal pears assorted)   
..                                                ...   
81  (blue 3 piece mini dots cutlery set, green 3 p...   
82  (pink 3 piece mini dots cutlery set, green 3 p...   
83     (key fob , back door , key fob , front  door )   
84  (poppy's playhouse livingroom, poppy's playhou...   
85  (lunch bag pink retrospot, lunch bag suki  des...   

                             consequents  antecedent support  \
0   (white hanging heart t-light holder)            0.058784   
1       (strawberry ceramic trinket box)            0.049254   
2                  (bathroom metal sign)            0.026993   
3         (red retrospot small milk jug)            0.023660   
4        (assorted colour bird ornament)            0.021927   
..                                   ...                 ...   
81  (pink 3 piece mini dots cutlery set)            0.011464   
82  (blue 3 piece mini dots cutlery set)            0.011997   
83                      (key fob , shed)            0.012463   
84          (poppy's playhouse bathroom)            0.013730   
85                (lunch bag red spotty)            0.013796   

    consequent support   support  confidence       lift  leverage  conviction  \
0             0.177953  0.042455    0.722222   4.058510  0.031995    2.959371   
1             0.074980  0.037457    0.760487  10.142533  0.033764    3.862089   
2             0.040589  0.021728    0.804938  19.831353  0.020632    4.918499   
3             0.037123  0.016796    0.709859  19.121592  0.015917    3.318652   
4             0.078512  0.016596    0.756839   9.639738  0.014874    3.789618   
..                 ...       ...         ...        ...       ...         ...   
81            0.029992  0.010064    0.877907  29.271370  0.009720    7.944827   
82            0.030525  0.010064    0.838889  27.481853  0.009698    6.017430   
83            0.025060  0.010064    0.807487  32.222153  0.009752    5.064272   
84            0.010997  0.010064    0.733010  66.655016  0.009913    3.704266   
85            0.075913  0.010064    0.729469   9.609260  0.009017    3.415821   

    zhangs_metric  
0        0.800671  
1        0.948103  
2        0.975918  
3        0.970669  
4        0.916356  
..            ...  
81       0.977037  
82       0.975313  
83       0.981194  
84       0.998709  
85       0.908467  

[86 rows x 10 columns]

# Analysis Results

From the market basket analysis, we can conclude that:
- The product Red Hanging Heart T-Light Holder was purchased in 5.9% of the total transactions.
- Once a buyer purchases Red Hanging Heart T-Light Holder, 72% of them will purchase White Hanging Heart T-Light Holder as well.
- There is a strong relationship between the Red Hanging Heart T-Ligh Holder and the White Hanging Heart T-Ligh Holder, with the chance of both products being purchased together in a single transaction being four times greater than the chance of both products being purchased individually.
- Based on this insight, the White Hanging Heart T-Ligh Holder product can be placed near/next to the Red Hanging Heart T-Ligh Holder or both products can be sold in bundles.